File: OCEL.CSharp.Tests/XmlTests.cs
using System;
using System.IO;
using Xunit;
using Xunit.Abstractions;

namespace OCEL.CSharp.Tests
{
    public class XmlTests
    {
        private readonly ITestOutputHelper _output;
        private readonly string _dataPath = Path.GetFullPath(Path.Combine("..", "..", "..", "..", "..", "..", "data", "OCEL"));

        public XmlTests(ITestOutputHelper output)
        {
            _output = output;
        }

        [Fact]
        public void SampleXmlIsValidAccordingToSchema()
        {
            var xml = File.ReadAllText(Path.Combine(_dataPath, "minimal.xmlocel"));
            Assert.True(OcelXml.Validate(xml));
        }

        [Fact]
        public void GitHubPm4PyXmlIsValidAccordingToSchema()
        {
            var xml = File.ReadAllText(Path.Combine(_dataPath, "github_pm4py.xmlocel"));
            Assert.True(OcelXml.Validate(xml));
        }

        [Fact]
        public void CanParseSampleXml()
        {
            var xml = File.ReadAllText(Path.Combine(_dataPath, "minimal.xmlocel"));
            var parsed = OcelXml.Deserialize(xml, true);
            Assert.NotNull(parsed);
        }

        [Fact]
        public void ParsedSampleXmlSatisfiedWellFormednessProperty()
        {
            var xml = File.ReadAllText(Path.Combine(_dataPath, "minimal.xmlocel"));
            var parsed = OcelXml.Deserialize(xml, true);
            Assert.True(parsed.IsValid);
        }

        [Fact]
        public void CanParseGitHubPm4PyLog()
        {
            var xml = File.ReadAllText(Path.Combine(_dataPath, "github_pm4py.xmlocel"));
            var parsed = OcelXml.Deserialize(xml, true);
            Assert.True(parsed.IsValid);
        }

        [Fact]
        public void ParsedGitHubPm4PyXmlSatisfiedWellFormednessProperty()
        {
            var xml = File.ReadAllText(Path.Combine(_dataPath, "github_pm4py.xmlocel"));
            var parsed = OcelXml.Deserialize(xml, true);
            Assert.True(parsed.IsValid);
        }

        [Fact]
        public void CanParseNestedXml()
        {
            var xml = File.ReadAllText(Path.Combine(_dataPath, "minimal_nested.xmlocel"));
            var parsed = OcelXml.Deserialize(xml, true);
            Assert.True(parsed.IsValid);
        }

        [Fact]
        public void CanSerializeNestedXml()
        {
            var xml = File.ReadAllText(Path.Combine(_dataPath, "minimal_nested.xmlocel"));
            var parsed = OcelXml.Deserialize(xml, true);
            var serialized = OcelXml.Serialize(parsed, Types.Formatting.Indented, true);
            _output.WriteLine($"Serialized XML:{Environment.NewLine}{serialized}");
            Assert.False(string.IsNullOrWhiteSpace(serialized));
        }

        [Fact]
        public void CanSerializeSampleLog()
        {
            var xml = File.ReadAllText(Path.Combine(_dataPath, "minimal.xmlocel"));
            var parsed = OcelXml.Deserialize(xml, true);
            var serialized = OcelXml.Serialize(parsed, Types.Formatting.Indented, true);
            _output.WriteLine($"Serialized XML:{Environment.NewLine}{serialized}");
            Assert.False(string.IsNullOrWhiteSpace(serialized));
        }

        [Fact]
        public void CanSerializeGitHubPm4PyLog()
        {
            var xml = File.ReadAllText(Path.Combine(_dataPath, "github_pm4py.xmlocel"));
            var parsed = OcelXml.Deserialize(xml, true);
            var serialized = OcelXml.Serialize(parsed, Types.Formatting.Indented, true);
            _output.WriteLine($"Serialized XML:{Environment.NewLine}{serialized}");
            Assert.False(string.IsNullOrWhiteSpace(serialized));
        }

        [Fact]
        public void CanSerializeSampleLogAndDeserializeItAgain()
        {
            var xml = File.ReadAllText(Path.Combine(_dataPath, "minimal.xmlocel"));
            var parsed = OcelXml.Deserialize(xml, true);
            var serialized = OcelXml.Serialize(parsed, Types.Formatting.Indented, true);
            var reSerialized = OcelXml.Deserialize(serialized, true);
            Assert.True(reSerialized.IsValid);
        }
    }
}

File: OCEL.Tests/XmlTests.fs
namespace OCEL.Tests

open OCEL
open OCEL.Types

open System
open System.IO
open Xunit
open Xunit.Abstractions

type XmlTests(output: ITestOutputHelper) =

    [<Fact>]
    member _.``Sample XML is valid according to schema`` () =
        let xml = File.ReadAllText("minimal.xmlocel")
        xml |> OcelXml.validate |> Assert.True

    [<Fact>]
    member _.``"GitHub pm4py" XML is valid according to schema`` () =
        let xml = File.ReadAllText("github_pm4py.xmlocel")
        xml |> OcelXml.validate |> Assert.True

    [<Fact>]
    member _.``Can parse sample XML`` () =
        let xml = File.ReadAllText("minimal.xmlocel")
        xml |> OcelXml.deserialize |> Assert.NotNull

    [<Fact>]
    member _.``Parsed sample XML satisfies well-formedness property`` () =
        let xml = File.ReadAllText("minimal.xmlocel")
        let parsed = OcelXml.deserialize xml
        parsed.IsValid |> Assert.True

    [<Fact>]
    member _.``Can parse "GitHub pm4py" log`` () =
        let xml = File.ReadAllText("github_pm4py.xmlocel")
        xml |> OcelXml.deserialize |> Assert.NotNull

    [<Fact>]
    member _.``Parsed "GitHub pm4py" XML satisfies well-formedness property`` () =
        let xml = File.ReadAllText("github_pm4py.xmlocel")
        let parsed = OcelXml.deserialize xml
        parsed.IsValid |> Assert.True

    [<Fact>]
    member _.``Can parse nested XML`` () =
        let xml = File.ReadAllText("minimal_nested.xmlocel")
        let parsed = OcelXml.deserialize xml
        parsed.IsValid |> Assert.True

    [<Fact>]
    member _.``Can serialize nested XML`` () =
        let xml = File.ReadAllText("minimal_nested.xmlocel")
        let parsed = OcelXml.deserialize xml
        let serialized = OcelXml.serialize Formatting.Indented parsed
        output.WriteLine $"Serialized XML:{Environment.NewLine}{serialized}"
        String.IsNullOrWhiteSpace serialized |> Assert.False

    [<Fact>]
    member _.``Can serialize sample log`` () =
        let xml = File.ReadAllText("minimal.xmlocel")
        let parsed = xml |> OcelXml.deserialize
        let serialized = parsed |> OcelXml.serialize Formatting.Indented
        output.WriteLine $"Serialized XML:{Environment.NewLine}{serialized}"
        String.IsNullOrWhiteSpace serialized |> Assert.False

    [<Fact>]
    member _.``Can serialize "GitHub pm4pmy" log`` () =
        let xml = File.ReadAllText("github_pm4py.xmlocel")
        let parsed = xml |> OcelXml.deserialize
        let serialized = parsed |> OcelXml.serialize Formatting.Indented
        output.WriteLine $"Serialized XML:{Environment.NewLine}{serialized}"
        String.IsNullOrWhiteSpace serialized |> Assert.False

    [<Fact>]
    member _.``Can serialize sample log and deserialize it again`` () =
        let xml = File.ReadAllText("minimal.xmlocel")
        let log = xml |> OcelXml.deserialize
        let logReserialized = log |> OcelXml.serialize Formatting.Indented |> OcelXml.deserialize
        log = logReserialized |> Assert.True
